syntax = "proto3";

package Db;

import "google/protobuf/descriptor.proto";

message DbSqlOptions {
}

message DbCacheOptions {
}

// 通用数据库选项
message FileOptionsDb {
    DbSqlOptions sql = 1;
    DbCacheOptions cache = 2;
    string name = 3;      // 数据库名称
}

extend google.protobuf.FileOptions {
    FileOptionsDb db = 20001;
}

// SQL 相关选项（MySQL）
message ForeignKey {
  string column = 1;             // 外键列
  string referenced_table = 2;   // 被引用表名
  string referenced_column = 3;  // 被引用列名
  string on_delete = 4;          // 外键删除策略 (e.g., CASCADE, RESTRICT, etc.)
  string on_update = 5;          // 外键更新策略 (e.g., CASCADE, RESTRICT, etc.)
}

message TableSqlOptions {
    string charset = 1;              // 字符集
    string engine = 2;               // 存储引擎
    // repeated int32 composite_key = 3;   // 复合主键 field_number
    // repeated ForeignKey foreign_keys = 4; // 外键定义
}

// Cache 相关选项（Redis）
message TableCacheOptions {
    int32 ttl = 1;                   // 键的 TTL（过期时间，秒）
}

message MessageOptionsTable {
    TableSqlOptions sql = 1;
    TableCacheOptions cache = 2;
    string name = 3;  // 表名
    int32 tick_second = 4;  // Tick入库时间
    int32 primary_key = 5;  // 主键 field_number
}

extend google.protobuf.MessageOptions {
    MessageOptionsTable table = 20001;
}

// SQL 相关字段选项（MySQL）
message ColumnSqlOptions {
    bool index = 1;                  // 索引
    bool unique = 2;                 // 唯一约束
    bool auto_increment = 3;         // 自增
    string default_value = 4;        // 默认值
    string comment = 5;              // 字段注释
    bool not_null = 6;               // 非空约束
}

// Cache 相关字段选项（Redis）
message ColumnCacheOptions {
    bool index = 1;                  // 字段是否作为索引
}

message FieldOptionsColumn {
    ColumnSqlOptions sql = 1;
    ColumnCacheOptions cache = 2;
}

extend google.protobuf.FieldOptions {
    FieldOptionsColumn column = 20001;
}
